### PR TITLE
Upgrade to NuGet.* v4.0.0 dependencies

### PR DESCRIPTION
**Pull request by xavierdecoster**
This PR

* upgrades our `NuGet.*` dependencies to version **4.0.0**
* consolidates `Moq` and `Newtonsoft.Json` dependencies to vLatest
* leverages the now public field `NuGet.Packaging.PackageIdValidator.MaxPackageIdLength` instead of copy/paste + [#sadpanda comment](https://github.com/NuGet/NuGetGallery/compare/dev...dev-pkgid-validation?expand=1#diff-0e2dd4dc7cd7407bf362381962b58326L17) in the gallery code base

This PR will also facilitate other validation improvements that are required for semver2 work (and didn't want to pollute those PRs with this upgrade).

### CODE REVIEW

**Comment by skofman1** [comment]
I have a PR changing gallery to .NET 4.6.2.. the merge conflicts I'm going to get break my ❤️

File: src/NuGetGallery.Core/Packaging/ManifestValidator.cs [hunk 1]
```diff
@@ -46,7 +43,7 @@ private static IEnumerable<ValidationResult> ValidateCore(PackageMetadata packag
             }
             else
             {
-                if (packageMetadata.Id.Length > MaxPackageIdLength)
+                if (packageMetadata.Id.Length > NuGet.Packaging.PackageIdValidator.MaxPackageIdLength)
```
Comment (scottbommarito) [inline src/NuGetGallery.Core/Packaging/ManifestValidator.cs]: One thing to note--(last time I checked) the gallery accepts `nupkg`s with up to a 128 characters in the id. Probably good to switch this to use this here but I question why we have this validating against 100 characters when the gallery accepts 128.

File: tests/NuGetGallery.Core.Facts/app.config [hunk 2]
```diff
@@ -1,9 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <configuration>
   <configSections>
-    <!-- For more information on Entity Framework configuration, visit http://go.microsoft.com/fwlink/?LinkID=237468 -->
+    
     <section name="entityFramework" type="System.Data.Entity.Internal.ConfigFile.EntityFrameworkSection, EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" requirePermission="false" />
-  </configSections>
+  <!-- For more information on Entity Framework configuration, visit http://go.microsoft.com/fwlink/?LinkID=237468 --></configSections>
```
Comment (scottbommarito) [inline tests/NuGetGallery.Core.Facts/app.config]: Did this comment get moved automatically?

Huge nit but I prefer where it was before.

Comment (xavierdecoster) [inline tests/NuGetGallery.Core.Facts/app.config]: Yes, nuget did this 🙌

File: src/NuGetGallery.Core/Packaging/ManifestValidator.cs (hunk 1 above)
Comment (xavierdecoster) [inline src/NuGetGallery.Core/Packaging/ManifestValidator.cs]: @scottbommarito any pointers to where in the code base we accept 128 chars? I did not change the value in this PR, so perhaps we should log a separate issue to track that one.

File: tests/NuGetGallery.Core.Facts/app.config (hunk 2 above)
Comment (xavierdecoster) [inline tests/NuGetGallery.Core.Facts/app.config]: Reverted this auto-change in https://github.com/NuGet/NuGetGallery/pull/3643/commits/4b624217bd4099354e69c9333c12924a384a20c1

File: src/NuGetGallery.Core/Packaging/ManifestValidator.cs (hunk 1 above)
Comment (scottbommarito) [inline src/NuGetGallery.Core/Packaging/ManifestValidator.cs]: https://github.com/NuGet/NuGetGallery/blob/0c6458164ec897c810baa6cd5f5e388c25a0ebd3/src/NuGetGallery.Core/CoreConstants.cs#L8

I don't know if you need to change this, but using the constant was part of https://github.com/NuGet/NuGetGallery/pull/3305 which was reverted.

Comment (xavierdecoster) [inline src/NuGetGallery.Core/Packaging/ManifestValidator.cs]: Interesting. I think we should definitely look into that some day. If I understand correctly, the gallery db is more forgiving than the server and client validation (the other way around would 've been worse though). Will leave it in as part of this PR, as I believe the server should leverage and align with the client validation.

Comment (xavierdecoster) [inline src/NuGetGallery.Core/Packaging/ManifestValidator.cs]: Tracked in issue https://github.com/NuGet/NuGetGallery/issues/3647

**Pull request merged**